benchmarks/test_web_client.py

```diff
@@ -0,0 +1,36 @@
+"""Benchmarks for C{HTTP11ClientProtocol}."""
+
+from twisted.internet.testing import StringTransport
+from twisted.web._newclient import HTTP11ClientProtocol, Request
+from twisted.web.http_headers import Headers
+
+RESPONSE = """HTTP/1.1 200 OK
+Host: blah
+Foo: bar
+Gaz: baz
+Content-length: 3
+
+abc""".replace(
+    "\n", "\r\n"
+).encode(
+    "utf-8"
+)
+
+
+def test_http_client_small_response(benchmark):
+    """Measure the time to run a simple HTTP 1.1 client request."""
+
+    def go():
+        for _ in range(1000):
+            protocol = HTTP11ClientProtocol()
+            protocol.makeConnection(StringTransport())
+            request = Request(
+                b"GET", b"/foo/bar", Headers({b"Host": [b"example.com"]}), None
+            )
+            response = protocol.request(request)
+            protocol.dataReceived(RESPONSE)
+            result = []
+            response.addCallback(result.append)
+            assert result
+
+    benchmark(go)
```

Comment:
I was expecting that `benchmark()` already handles running the test multiple time... including having the test rounds as a parameter.

I expect there will be a lot of boilerplate code for  the tests like

```
def test_something(benchmark): 
    def go():
        for _ in range(1000):
          ACTUAL TEST
   benchmark(go)
```

Maybe we can have a decorator like this.

```
@generate_benchmark_test(__name__)
def http_client_small_response():
   """
   Docstring for this test.
   """
   protocol = HTTP11ClientProtocol()
   assert protocol
   MORE TEST CODE
```

It will "inject" into the module scope, new functions via
```setattr(sys.modules[module_name], name, new_test)```

name will start with `test_` to be picked up the pytest test discovery.

Not sure if this will work with pytest

I am using something like this with stdlib unittest


Other than fixture and parameters, I couldn't find any other way to dynamicaly generate tests in pytest